PyRoute/SystemData/UWP.py

```diff
@@ -0,0 +1,256 @@
+"""
+Created on Nov 21, 2023
+
+@author: CyberiaResurrection
+
+Along the lines of TradeCodes, pull all the UWP handling, checking, etc into one single class that _just_ does UWP,
+rather than the multiple concerns the Star class has evolved to embody.
+"""
+import re
+
+from PyRoute.SystemData.Utilities import Utilities
+
+
+class UWP(object):
+    # Port code, size, atmo, hydro, pop, gov, law, the all-important hyphen, then TL
+    match_string = '^([A-HXYa-hxy\?])([0-9A-Fa-f\?])([0-9A-Fa-f\?])([0-9Aa])([0-9A-Fa-f\?])([0-9A-Za-z\?])([0-9A-Ja-j\?])-([0-9A-Za-z\?])'
+
+    match = re.compile(match_string)
+
+    # various limits
+    flux = 5
+    atmo_limit = 15
+    hydro_limit = 10
+    hydro_atm_mod = -4
+    gov_limit = 15
+    law_limit = 18
+    tl_limit = 33
+
+    def __init__(self, uwp_line):
+        matches = UWP.match.match(uwp_line)
+        if not matches:
+            raise ValueError('Input UWP malformed')
+        self.line = str(matches[0]).upper()
+        self.port = self.line[0]
+        self.size = self.line[1]
+        self.atmo = self.line[2]
+        self.hydro = self.line[3]
+        self.pop = self.line[4]
+        self.gov = self.line[5]
+        self.law = self.line[6]
+        self.tl = self.line[8]
+        self._size_code = self._ehex_to_int(self.size)
+        self._atmo_code = self._ehex_to_int(self.atmo)
+        self._hydro_code = self._ehex_to_int(self.hydro)
+        self._pop_code = self._ehex_to_int(self.pop)
+        self._gov_code = self._ehex_to_int(self.gov)
+        self._law_code = self._ehex_to_int(self.law)
+        self._tl_code = self._ehex_to_int(self.tl)
+
+    def __str__(self):
```

Comment:
Please add a:

```
def __repr__(self):
   return self.line
```
This is for unicode support. There isn't any need for unicode support in the UWP itself, but there is just enough floating in the Traveller Map data that we'd like to make sure it isn't going to cause problem.s

Reply:
Pushed up